src/DurableTask.Netherite/OrchestrationService/NetheriteOrchestrationService.cs

```diff
@@ -726,24 +758,192 @@ async Task<PurgeResult> IOrchestrationServicePurgeClient.PurgeInstanceStateAsync
+                    FetchInput = filter.IncludeState,
+                    PrefetchHistory = false,
+                    ExcludeEntities = false,
+                }; 
+
+                List<EntityMetadata> metadataList = new List<EntityMetadata>();
+
+                InstanceQueryResult result = await (await this.service.GetClientAsync().ConfigureAwait(false))
+                    .QueryOrchestrationStatesAsync(condition, filter.PageSize ?? 200, filter.ContinuationToken, cancellation).ConfigureAwait(false);
+
+                foreach(var entry in result.Instances)
+                {
+                    var metadata = this.GetEntityMetadata(entry, filter.IncludeState, filter.IncludeTransient);
+                    if (metadata.HasValue)
+                    {
+                        metadataList.Add(metadata.Value);
+                    }
+                }
+
+                return new EntityQueryResult()
+                {
+                    Results = metadataList,
+                    ContinuationToken = result.ContinuationToken,
+                };
+            }
+
+            public override async Task<CleanEntityStorageResult> CleanEntityStorageAsync(CleanEntityStorageRequest request = default, CancellationToken cancellation = default)
+            {
+                if (!request.ReleaseOrphanedLocks)
+                {
+                    // there is no need to do anything since deletion is implicit
+                    return new CleanEntityStorageResult();
+                }
+
+                var condition = new InstanceQuery()
+                {
+                    InstanceIdPrefix = "@",
+                    FetchInput = false,
+                    PrefetchHistory = false,
+                    ExcludeEntities = false,
+                };
+
+                var client = await this.service.GetClientAsync().ConfigureAwait(false);
+
+                string continuationToken = null;
+                int orphanedLocksReleased = 0;
+                            
+                // list all entities (without fetching the input) and for each one that requires action,
+                // perform that action. Waits for all actions to finish after each page.
+                do
+                {
+                    var page = await client.QueryOrchestrationStatesAsync(condition, 500, continuationToken, cancellation).ConfigureAwait(false);
+
+                    List<Task> tasks = new List<Task>();
+                    foreach (var state in page.Instances)
+                    {
+                        EntityStatus status = ClientEntityHelpers.GetEntityStatus(state.Status);
+                        if (status != null)
+                        {
+                            if (request.ReleaseOrphanedLocks && status.LockedBy != null)
+                            {
+                                tasks.Add(CheckForOrphanedLockAndFixIt(state, status.LockedBy));
+                            }
```

Comment:
have we always had this functionality of releasing orphaned locks? Maybe I missed it.

Reply:
Not "always", but it has been there for a long time now.